web/middleware/csrf.go

```diff
@@ -0,0 +1,45 @@
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/bashmohandes/go-askme/web/framework"
+	"github.com/google/uuid"
+)
+
+// CSRF adds a csrf token to all sessions
+type csrf struct {
+}
+
+// Add CSRF token if it does not exist in session
+func (c *csrf) Run(cxt framework.Context) bool {
+	session := cxt.Session()
+	if session.Get("csrf") == nil {
+		session.Set("csrf", uuid.New().String())
+	}
+	return true
+}
+
+func CSRF() framework.Middleware {
+	return &csrf{}
+}
+
+// Decorator for framework.RouteHandler that checks the
+// validity of the CSRF token by comparing the token in the
+// form (hidden input) and the correct token in the session.
+// It gives 400 Bad Request error if it is invalid and calls the
+// given RouteHandler otherwise.
+// E.g., router.Post("/login", middleware.RequireCSRF(performLogin))
+// where performLogin is of type framework.RouteHandler
+func RequireCSRF(f framework.RouteHandler) framework.RouteHandler {
+	return func(cxt framework.Context) {
+		sessionCSRF := cxt.Session().Get("csrf")
+		formCSRF := cxt.Request().PostFormValue("csrf")
+		if sessionCSRF != formCSRF {
+			w := cxt.ResponseWriter()
+			http.Error(w, "400 - Bad Request!", http.StatusBadRequest)
```

Comment:
We need to make sure the response code is 403

Reply:
That was fixed [here](https://github.com/ZeyadYasser/go-askme/blob/2920053694fbf22c8d89dc4f0b637f936e99aea5/web/middleware/csrf/csrf.go#L49).